### PR TITLE
Add sync tests to mocha suite

### PR DESCRIPTION
**Pull request by stoically**
This adds all sync tests to the mocha suite. I've also refactored the tests first to not rely on "globals magic", which was a bad idea in the first place and hopefully helps with understanding the test flow. I'm happy to answer questions if anything is unclear. I had to [fix a small bug in `webextensions-api-fake`](https://github.com/stoically/webextensions-api-fake/commit/9a1bbf7f7d4105d486d0d2e00ed890816fce95b6) as well, so you might need to do a fresh `npm install` to have a successful test run.

- `npm run test` lints and generates a coverage report in the `coverage` directory, where you could open the `index.html` to check which branches in the `sync.js` are covered by the test
- `npm run test:once` runs the tests once without linting and coverage
- `npm run test:watch` watches the tests

To only execute one specific test / suite it's possible to change `it` to `it.only` or `describe` to `describe.only` respectively.

What's missing in this PR is changing the `console.assert` calls to `should` or `expect` assertions. Would you be interested in giving that a go or should I look into that?

Something I'd like to do in a follow-up if we can get this PR merged is adding a test that simulates sync between multiple browsers using the high-level APIs without the test knowing about the inner structure of the storage.

### CODE REVIEW

File: test/features/sync.test.js [hunk 1]
```diff
@@ -1,64 +1,695 @@
...
+
+    const macConfigs = await this.webExt.browser.storage.local.get();
+    const identities = [];
+    for(const configKey of Object.keys(macConfigs)) {
+      if (configKey.includes("identitiesState@@_") && !configKey.includes("default")) {
+        identities.push(macConfigs[configKey]);
+      }
+    }
+
+    console.assert(identities.length === 5, "There should be 5 identity entries");
```
Comment (stoically) [inline test/features/sync.test.js:37]: Here an example how asserting with `should` could look like:

```suggestion
    identities.should.have.lengthOf(5);
    for (const identity of identities) {
      identity.macAddonUUID.should.be.ok;
    }
```

[chai reference](https://www.chaijs.com/api/bdd/)

**Comment by kendallwerts** [comment]
Thanks for this! It looks great, but after I upgrade my npm packages, I get a syntax error when I run any of the mocha tests. Can you commit your `package-lock.json` file, maybe? I'm not sure what I've got that's not aligned.

**Comment by kendallwerts** [comment]
Well, it doesn't look like it's your stuff. I can't get `npm test` to run properly on any version--not even master

**Comment by stoically** [comment]
What's the error you seeing? And what platform and which npm/node version are you running?

File: test/features/sync.test.js [hunk 2]
```diff
@@ -1,64 +1,695 @@
...
+    await this.webExt.browser.storage.onChanged.removeListener(this.webExt.background.window.sync.storageArea.onChangedListener);
+    await this.webExt.background.window.sync.removeContextualIdentityListeners();
+  }
+  
+  async setState(syncData, localData, identityData, assignmentData){
+    await this.removeAllContainers();
+    await this.webExt.browser.storage.sync.clear();
+    await this.webExt.browser.storage.sync.set(syncData);
+    await this.webExt.browser.storage.local.clear();
+    await this.webExt.browser.storage.local.set(localData);
```
Comment (kendallwerts) [inline test/features/sync.test.js:304]: Awesome. This is what I was having the most trouble with.

**Comment by kendallwerts** [comment]
> What's the error you seeing? And what platform and which npm/node version are you running?

I got it working. I guess I had a `node` installed and a `nodejs` installed. `node` was version 8, and it started breaking when mocha ran. Not sure what changed to make it start using `node` instead of `nodejs`, but I fixed it up.

File: test/features/sync.test.js (hunk 2 above)
Comment (stoically) [reply test/features/sync.test.js:304]: Thanks for merging. Just realized that I missed some of the comments from the original sync test file, sorry about that.

Comment (kendallwerts) [reply test/features/sync.test.js:304]: Not sure what you mean, but I'm going over it all.

Comment (stoically) [reply test/features/sync.test.js:304]: e.g. https://github.com/kendallcorner/multi-account-containers/blob/e6f5bf76c2e501d424a1036a580ca0b16abc3ae4/src/js/background/test.js#L308-L315 is what I meant.

Comment (stoically) [reply test/features/sync.test.js:304]: > but I'm going over it all.

Great! I'll look into the "browser simulation sync" in the coming days.

If you want to give it a shot, it should work along the lines of:

```js
const webext1 = await initializeWithTab();
const webext2 = await initializeWithTab();

webext1OrigSyncSet = webext1.browser.storage.sync.set;
webext2OrigSyncSet = webext2.browser.storage.sync.set;
webext1OrigSyncRemove = webext1.browser.storage.sync.remove;
webext2OrigSyncRemove = webext2.browser.storage.sync.remove;

webext1.browser.storage.sync.set.callsFake((...args) => {
  webext1OrigSyncSet(...args);
  webext2OrigSyncSet(...args);
  webext2.browser.storage.onChanged.addListener.yield(null, "sync");
});

webext1.browser.storage.sync.remove.callsFake((...args) => {
  webext1OrigSyncRemove(...args);
  webext2OrigSyncRemove(...args);
  webext2.browser.storage.onChanged.addListener.yield(null, "sync");
});

webext2.browser.storage.sync.set.callsFake((...args) => {
  webext2OrigSyncSet(...args);
  webext1OrigSyncSet(...args);
  webext1.browser.storage.onChanged.addListener.yield(null, "sync");
});

webext2.browser.storage.sync.remove.callsFake((...args) => {
  webext2OrigSyncRemove(...args);
  webext1OrigSyncRemove(...args);
  webext1.browser.storage.onChanged.addListener.yield(null, "sync");
});
```

This is a bit verbose, but basically just wires the sync.set&remove calls up between two webextension instances - so calls to e.g.

```js
await webext1.browser.contextualIdentities.create({name: "foo"});
```

should fire the

```js
webext2.browser.storage.onChanged
```

event and hence

```js
await webext2.browser.contextualIdentities.query({});
```

should return the synced container accordingly.

**Pull request merged**